ludwig/trainers/trainer.py

```diff
@@ -353,115 +365,73 @@ def train_for_tuning(
-                    else:
-                        if smoothed_loss < best_loss or step_count == 0:
-                            best_loss = smoothed_loss
-
-                    # compute new learning rate
-                    if mode == "exponential":
-                        current_learning_rate = exponential_scheduler(current_learning_rate, step_count)
-                    else:
-                        current_learning_rate = linear_scheduler(current_learning_rate, step_count)
-
-                    self.set_optimizer_learning_rate(current_learning_rate)
-                    step_count += 1
+    def tune_learning_rate(self, config: ModelConfigDict) -> float:
+        """Recommend a learning rate based on the feautres in the configs and the associated encoders.
 
-                epoch += 1
+        param:
+            config: Ludwig config used to train the model.
+        """
+        logger.info("Tuning learning rate...")
 
-        optimal_lr = get_optimal_lr(losses, learning_rates)
-        if optimal_lr:
-            learning_rate = optimal_lr
+        features_with_large_encoders = {TEXT, IMAGE, SEQUENCE, AUTO, TIMESERIES}
+        huggingface_encoders = {
+            "vit",
+            "ctrl",
+            "camembert",
+            "xlnet",
+            "distilbert",
+            "falubert",
+            "transformer_xl",
+            "xlm",
+            "gpt",
+            "gpt2",
+            "bert",
+            "roberta",
+            "electra",
+            "longformer",
+            "t5",
+            "albert",
+            "mt5",
+            "xlmroberta",
+        }
+
+        encoders_in_config = set()
+        features_in_config = set()
+        trainable_in_config = set()
+
+        # input and output feature sections
+        for feature in config["input_features"] + config["output_features"]:
+            features_in_config.add(feature.get("type"))
+            feature_encoder = feature.get("encoder", {})
+            encoder_type = feature_encoder.get("type")
+            encoder_is_trainable = feature_encoder.get("trainable")
+            trainable_in_config.add(encoder_is_trainable)
+            encoders_in_config.add(encoder_type)
+
+        # default section
+        defaults = config.get("defaults", {})
+        for feature_type in defaults.keys() & features_in_config:
+            feature_defaults = defaults.get(feature_type, {})
+            encoder_section = feature_defaults.get("encoder", {})
+            encoder_type = encoder_section.get("type")
+            encoder_is_trainable = encoder_section.get("trainable")
+            trainable_in_config.add(encoder_is_trainable)
+            encoders_in_config.add(encoder_type)
+
+        is_feature_with_large_encoder = len(features_in_config & features_with_large_encoders) > 0
+        has_huggingface_encoder = len(encoders_in_config & huggingface_encoders) > 0
+        has_trainable_encoder = any(trainable_in_config)
+
+        if is_feature_with_large_encoder:
```

Comment:
Can you add some comments explaining the intuition or data backing these heuristics? At a glance they feel arbitrary, but I assume there's a reason these specific values were chosen.

Reply:
Comment added